Mk nikshay followup api

mkangia opened the pull request.

mkangia committed:
add field_mappings, utils, view, url and setting

mkangia committed:
Add Follow Up generator

mkangia committed:
add follow up repeater

mkangia committed:
add tests for Follow Up generator and repeater

dimagimon commented:
There are 7 errors:

* custom/enikshay/integrations/nikshay/repeater_generator.py, line 293 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 294 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 308 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 640 - E303 too many blank lines (2)
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 817 - E121 continuation line under-indented for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 842 - E123 closing bracket does not match indentation of opening bracket's line
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 911 - W292 no newline at end of file

mkangia committed:
fix failing tests due to changes for max_bacilli_count and testing_facility

dimagimon commented:
There are 8 errors:

* custom/enikshay/integrations/nikshay/repeater_generator.py, line 294 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 295 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 309 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 640 - E303 too many blank lines (2)
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 818 - E121 continuation line under-indented for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 843 - E123 closing bracket does not match indentation of opening bracket's line
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 907 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 910 - W292 no newline at end of file

mkangia committed:
remove check for nikshay registered

to enable sending updates for migrated patients or patients with nikshay id

mkangia committed:
fix condition for test submission check

mkangia committed:
fetch lab referral for dmc_location in case not found on testing_facility_id

this is fallback for some mobile users, if any on older version of app

mkangia committed:
revert incorrect fix for test submission check and fix tests due to change in fallback for dmc code

dimagimon commented:
There are 7 errors:

* custom/enikshay/integrations/nikshay/repeater_generator.py, line 294 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 295 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 313 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 817 - E121 continuation line under-indented for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 842 - E123 closing bracket does not match indentation of opening bracket's line
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 910 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 913 - W292 no newline at end of file

proteusvacuum commented on custom/enikshay/exceptions.py:
```diff
@@ -12,3 +12,7 @@ class NikshayLocationNotFound(ENikshayException):
     pass
+
+
+class RequiredValueMissing(ENikshayException):
```
nit: Might be clearer if this was called `NikshayRequiredValueMissing`, so that it fits the pattern of the rest of the exceptions

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+            "LabNo": lab_serial_number,
+            "Local_ID": person_case.get_id,
+            "IntervalId": interval_id,
+            # since weight is not taken and is mandatory we send 1
```
I think we collect weight in the treatment card form. It is stored on the `episode`. You can check it out here: https://india.commcarehq.org/a/enikshay/apps/view/ce3e9160e05acd8e0020ed5e33ca89ea/summary/#/cases

mkangia replied:
The reason we have this as "1" because weight is not collected on each follow up so it was decided to not mention it at all instead of having the same value from episode each time and since its mandatory we send 1.

proteusvacuum replied:
👍

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+        test_case_properties = test_case.dynamic_case_properties()
+        episode_case_properties = episode_case.dynamic_case_properties()
+
+        interval_id, lab_serial_number, result_grade, dmc_code = self._get_mandatory_fields(test_case_properties)
```
I think you could do 
```py
properties_dict.update(self._get_mandatory_fields(...))` 
```
and have that method return a dict.

mkangia replied:
i think i did that because i thought it was good to have all payload specific properties being set in a single place to look at them all at once. So its fine with base properties being set separately since they are not payload specific. Just for better view.

proteusvacuum replied:
I still think this is somewhat overly verbose and will be annoying to change if there are extra "mandatory properties" later, but I guess it's fine if you want to leave like this.

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+                                            test_case_properties.get('follow_up_test_reason'))
+
+        dmc_code = self._get_dmc_code(test_case_properties)
+        lab_serial_number = test_case_properties.get('lab_serial_number', None)
```
nit: `None` is the default for the `get` method.

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+        bacilli_count = test_case_properties.get('max_bacilli_count', None)
+        result_grade = self.get_result_grade(test_result_grade, bacilli_count)
+
+        if any(mandatory_value is None for mandatory_value in [lab_serial_number, result_grade]):
```
does this need to check that any of the values are 'blank' (i.e. `''`)? If not, you could just do:
```py
try:
    test_case_properties['result_grade']
    test_case_properties['lab_serial_number']
except AttributeError:
    raise RequiredValueMissing(...)
```
If you do need to check blank values, I think you can just do:
```py
if lab_serial_number and result_grade:
    ...
```

mkangia replied:
Oh yes.... thanks for raising this Farid. Updating to check for blank string as well.

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+                    .format(location_id=dmc_location_id, test_case_id=self.test_case.get_id)
+            )
+        nikshay_code = dmc.metadata.get('nikshay_code')
+        if not nikshay_code or (isinstance(nikshay_code, (str, unicode)) and not nikshay_code.isdigit()):
```
you can use `basestring` here instead

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+            )
+        nikshay_code = dmc.metadata.get('nikshay_code')
+        if not nikshay_code or (isinstance(nikshay_code, (str, unicode)) and not nikshay_code.isdigit()):
+            raise RequiredValueMissing("InAppt value for dmc, got value: {}".format(nikshay_code))
```
what does `InAppt` mean?

mkangia replied:
Inappropriate. The code should be a string with just integer values. I think QA and test locations would keep nikshay code like "QA123" which eventually fails at Nikshay since it expects to parse it as integer.

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+        return dmc.metadata.get('nikshay_code')
+
+    def handle_success(self, response, payload_doc, repeat_record):
+        # Simple success message that has {"Nikshay_Message": "Success"...}
```
I don't think this comment is true. This method runs when `200 <= response.status < 300`

mkangia replied:
Ya thats alright. Its just that the message has this content. Will remove it, its confusing because failure messages also have Success in it. :(

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeaters.py:
```diff
@@ -128,6 +134,49 @@ def allowed_to_forward(self, episode_case):
+        app_label = 'repeaters'
+
+    include_app_id_param = False
+    friendly_name = _("Forward eNikshay Patient's Follow Ups to Nikshay")
```
can you add the expected case type to the friendly name

mkangia replied:
I have added case types to the index itself now [here](https://github.com/dimagi/commcare-hq/pull/15430/files#diff-a358d71a8a9e9a9295ee52c447c5f9b2R4). Should we still set it in the friendly name?

proteusvacuum replied:
Yes. I think it would be great to document in the UI which case types this expects (this is so that we don't have to override the `new` view to always have the right case type).

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeaters.py:
```diff
@@ -128,6 +134,49 @@ def allowed_to_forward(self, episode_case):
+            test_case_properties = test_case.dynamic_case_properties()
+            episode_case_properties = episode_case.dynamic_case_properties()
+            return (
+                date_tested_added_for_test(test_case) and
```
can you use the `case_properties_changed` method instead of this? 
Also, put this after the other conditions, because I think it does a bunch of db lookups.

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeaters.py:
```diff
@@ -128,6 +134,49 @@ def allowed_to_forward(self, episode_case):
+            episode_case_properties = episode_case.dynamic_case_properties()
+            return (
+                date_tested_added_for_test(test_case) and
+                test_case_properties.get('nikshay_registered', 'false') == 'false' and
```
this is going to default to `True`, is that what you expect? i.e.  if `nikshay_registered` is not present, this will pass

mkangia replied:
Yup! sent a follow up for test if has not been sent. so by default it should be true. If it has been sent already don't trigger. Does that not sound right?

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeaters.py:
```diff
@@ -157,9 +206,23 @@ def related_dates_changed(case):
     return value_changed
 
 
+def date_tested_added_for_test(case):
```
please use the helper method for this instead.

mkangia replied:
sorry these changes were done before that method was added. Updated now.

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+    def test_followup_for_tests(self):
+        self.assertEqual(NikshayFollowupRepeater().followup_for_tests, ['end_of_ip', 'end_of_cp'])
+
+    @run_with_all_backends
```
please only run this on the SQL backend.

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+    def test_trigger(self):
+        self.repeat_record_count = 0
+
+        def check_repeat_record_added():
```
nit: I think if you called this `assertRepeatRecordAdded` and raised an `AssertionError` it will result in better error messages if this test fails

proteusvacuum replied:
you could add that onto the `NikshayRepeaterTestBase`

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -88,6 +90,9 @@ def _create_nikshay_registered_case(self):
         )
         self.create_case(nikshay_registered_case)
 
+    def update_case_with(self, case_id, case_attrs):
```
elsewhere this is called `update_case`. Would be good to maintain that convention.

mkangia replied:
added this method to avoid calling update_case with domain. It calls `update_case` internally. Just makes things better to read and at times in a single line.

proteusvacuum replied:
I still think you should remove the word `with`, or change it to be `update_case_with_attributes` although that seems redundant.

mkangia replied:
removing the method and using update_case only

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+        self.assertEqual(payload["DMC"], '123')
+        self.assertEqual(payload["PatientID"], DUMMY_NIKSHAY_ID)
+
+    @run_with_all_backends
```
run only on SQL

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+        test_case = CaseAccessors(self.domain).get_case(self.test_id)
+        NikshayFollowupPayloadGenerator(None).get_payload(self.repeat_record, test_case)
+
+    def test_mandatory_field_dmc_code(self):
```
run on SQL

proteusvacuum commented on custom/enikshay/integrations/utils.py:
```diff
@@ -24,3 +27,20 @@ def is_valid_episode_submission(episode_case):
+            {lab_referral_id}"
+            .format(location_id=lab_referral_case.owner_id, lab_referral_id=lab_referral_case.case_id)
+        )
+    return dmc_location.metadata.get('is_test', "yes") == "yes"
```
👍

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+
+        self._create_nikshay_registered_case()
+
+        MockRepeater = namedtuple('MockRepeater', 'username password')
```
looks like this is copy/pasted from other tests. might be worth making it part of the super class?

mkangia replied:
just used in two setups and not in others. I think its fine to keep it like this.

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+        self.repeat_record = MockRepeatRecord(MockRepeater(username="arwen", password="Hadhafang"))
+
+    def create_case_structure(self):
+        return {case.get_id: case for case in filter(None, self.factory.create_or_update_cases(
```
what does the `filter` do here?

mkangia replied:
i have not looked into the reason behind this but `create_or_update_cases` returns few Nones from that method. so this filters those out.

proteusvacuum reviewed (changes requested):
Couple of comments. Great tests!
Also, there are conflicts with master.

mkangia commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -68,6 +75,7 @@ def _base_properties(self, repeat_record):
         return {
             "regBy": username,
             "regby": username,
+            "RegBy": username,
```
ya its different for some end point and i think same for two of the 4 end points. It would a dream come true if they make it a single representation of that. Have requested it.

proteusvacuum replied:
😭

mkangia commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -219,6 +227,117 @@ def handle_exception(self, exception, repeat_record):
+@RegisterGenerator(NikshayFollowupRepeater, 'case_json', 'JSON', is_default=True)
+class NikshayFollowupPayloadGenerator(BaseNikshayPayloadGenerator):
+    def get_payload(self, repeat_record, test_case):
+        self.test_case = test_case
```
well its accessed later in method calls.

proteusvacuum replied:
oh interesting, I missed that. I got a little confused because you've mixed instance variables and passing arguments into the private methods (`self.test_case`, `test_case_properties`). I would probably vote for just passing this into the methods that require it.

mkangia committed:
review fixes

RequiredValueMissing -> NikshayRequiredValueMissing
redundant None return from get
basestring instead of (str, unicode) to check for string
use case_properties_changed
use only sql backend in tests since nikshay is now on sql

mkangia committed:
fix failing test due to AccessRestricted on save with DisabledDbMixin

dimagimon commented:
There are 8 errors:

* custom/enikshay/integrations/nikshay/repeater_generator.py, line 292 - E501 line too long (117 > 115 characters)
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 294 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 295 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 313 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 816 - E121 continuation line under-indented for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 841 - E123 closing bracket does not match indentation of opening bracket's line
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 910 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 913 - W292 no newline at end of file

mkangia committed:
Merge master: Resolve conflicts

dimagimon commented:
There are 8 errors:

* custom/enikshay/integrations/nikshay/repeater_generator.py, line 292 - E501 line too long (117 > 115 characters)
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 294 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 295 - E122 continuation line missing indentation or outdented
* custom/enikshay/integrations/nikshay/repeater_generator.py, line 313 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 816 - E121 continuation line under-indented for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 841 - E123 closing bracket does not match indentation of opening bracket's line
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 910 - E131 continuation line unaligned for hanging indent
* custom/enikshay/integrations/nikshay/tests/test_repeaters.py, line 913 - W292 no newline at end of file

mkangia committed:
indentation fixes [skip ci]

dimagimon commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+        with self.assertRaisesMessage(NikshayRequiredValueMissing,
+                                      "Value missing for intervalID, purpose_of_testing: {testing_purpose}, "
+                                      "follow_up_test_reason: {follow_up_test_reason}".format(
+                                            testing_purpose="testing",
```
E126 continuation line over-indented for hanging indent

dimagimon commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+                NikshayRequiredValueMissing,
+                "Mandatory value missing in one of the following LabSerialNo: {lsn}, ResultGrade: {rg}".format(
+                    lsn=test_case.dynamic_case_properties().get('lab_serial_number'), rg="scanty")
+                ):
```
E123 closing bracket does not match indentation of opening bracket's line

dimagimon commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+                .format(location_id=123,
+                        test_case_id=self.test_id)
+        ):
+            NikshayFollowupPayloadGenerator(None).get_payload(self.repeat_record, test_case)
```
W292 no newline at end of file

mkangia committed:
fix indentation and eof [skip ci]

mkangia commented:
@proteusvacuum thanks for the review. Responded to comments and made required changes as well.

proteusvacuum commented on custom/enikshay/integrations/nikshay/tests/test_repeaters.py:
```diff
@@ -600,6 +600,7 @@ def test_payload_properties(self, _):
         self.assertEqual(payload['Outcome'], '7')
 
 
+@use_sql_backend
```
seems like this is a poorly named decorator. What you really want is: `@override_settings(TESTS_SHOULD_USE_SQL_BACKEND=True)`

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeater_generator.py:
```diff
@@ -313,12 +313,11 @@ def _get_dmc_code(self, test_case_properties):
-        if not nikshay_code or (isinstance(nikshay_code, (str, unicode)) and not nikshay_code.isdigit()):
-            raise RequiredValueMissing("InAppt value for dmc, got value: {}".format(nikshay_code))
+        if not nikshay_code or (isinstance(nikshay_code, basestring) and not nikshay_code.isdigit()):
+            raise NikshayRequiredValueMissing("InAppt value for dmc, got value: {}".format(nikshay_code))
```
Still not sure what `InAppt` means.

mkangia replied:
(offlined) need to change to Inappropriate.

proteusvacuum reviewed (approved):
Few non-blocking comments, looks great!

mkangia committed:
replace cpt_initiation_date with cpt_1_date for HIV test updates

mkangia committed:
Avoid breaking phone submissions in case of missing episode and occurrence cases

mkangia committed:
pass test_case as argument instead of setting on object

mkangia committed:
add expected case type to friendly name

mkangia committed:
InAppt -> Inappropriate

mkangia committed:
remove update_case_with method

mkangia committed:
use override settings instead of use_sql_backend since thats for testing with shards

mkangia commented:
@proteusvacuum Changes done. Would be good to see changes post last review,
Also sneaked in the fix for CPT delivery date for HIV Test repeater.

proteusvacuum commented on custom/enikshay/integrations/nikshay/repeaters.py:
```diff
@@ -200,7 +203,7 @@ def related_dates_changed(case):
         action for action in last_update_actions
```
Would be good to at some point change these to the `case_properties_changed` method so everything is consistent (can do in a follow-up PR (also made this comment on the PR that introduced this change)

proteusvacuum commented:
Thanks! Looks great!

mkangia committed:
Merge master: Resolve conflicts
